Sources/Interaction/Manipulators/SliceManipulator/index.js

```diff
@@ -0,0 +1,111 @@
+import macro from 'vtk.js/Sources/macro';
+import vtkCameraManipulator from 'vtk.js/Sources/Interaction/Manipulators/CameraManipulator';
+import vtkMath from 'vtk.js/Sources/Common/Core/Math';
+
+// ----------------------------------------------------------------------------
+// vtkSliceManipulator methods
+// ----------------------------------------------------------------------------
+
+function vtkSliceManipulator(publicAPI, model) {
+  // Set our className
+  model.classHierarchy.push('vtkSliceManipulator');
+
+  publicAPI.onAnimation = (interactor, renderer) => {
+    const lastPtr = interactor.getPointerIndex();
+    const pos = interactor.getAnimationEventPosition(lastPtr);
+    const lastPos = interactor.getLastAnimationEventPosition(lastPtr);
+
+    if (!pos || !lastPos || !renderer) {
+      return;
+    }
+
+    const dy = pos.y - lastPos.y;
+    console.log(dy);
+
+    const camera = renderer.getActiveCamera();
+    const range = camera.getClippingRange();
+    let distance = camera.getDistance();
+
+    // scale the interaction by the height of the viewport
+    let viewportHeight = 0.0;
+    if (camera.getParallelProjection()) {
+      viewportHeight = camera.getParallelScale();
+    } else {
+      const angle = vtkMath.radiansFromDegrees(camera.getViewAngle());
+      viewportHeight = 2.0 * distance * Math.tan(0.5 * angle);
+    }
+
+    const size = interactor.getView().getViewportSize(renderer);
+    const delta = dy * viewportHeight / size[1];
+    distance += delta;
+
+    // clamp the distance to the clipping range
+    if (distance < range[0]) {
+      distance = range[0] + viewportHeight * 1e-3;
+    }
+    if (distance > range[1]) {
+      distance = range[1] - viewportHeight * 1e-3;
+    }
+    camera.setDistance(distance);
+  };
+
+  publicAPI.onPinch = (interactor) => {
+    const interactorStyle = interactor.getInteractorStyle();
+    let renderer = interactorStyle.getCurrentRenderer();
+
+    if (!renderer) {
+      const pos = interactor.getAnimationEventPosition(
+        interactor.getPointerIndex()
+      );
+      renderer = interactor.findPokedRenderer(pos);
+      if (!renderer) {
+        return;
+      }
+    }
+
+    let delta = interactor.getScale() / interactor.getLastScale();
+    delta = 1.0 - delta;
+    delta *= 25; // expose factor?
+    console.log(delta);
+
+    const camera = renderer.getActiveCamera();
+    const range = camera.getClippingRange();
+    let distance = camera.getDistance();
+    distance += delta;
+
+    // clamp the distance to the clipping range
+    if (distance < range[0]) {
+      distance = range[0];
+    }
+    if (distance > range[1]) {
+      distance = range[1];
+    }
+    camera.setDistance(distance);
```

Comment:
@jourdain @floryst @martinken what are your thoughts on this?

This is the same as seen in [`vtkInteractorStyleImage`](https://github.com/Kitware/vtk-js/blob/b807950244e1ed7db1ae6858c6829ce33bc32786/Sources/Interaction/Style/InteractorStyleImage/index.js#L261-L268): this ends up calling [`vtkImageMapper.setZSliceFromCamera`](https://github.com/Kitware/vtk-js/blob/master/Sources/Rendering/Core/ImageMapper/index.js#L19-L25) where the focal point (defined here by `distance`) is converted into a slice index to display.

The issue I see is that this distance is clamped by the clipping range of the camera, which I assume should match the bounds of the volume. However, when testing in the `imageMapper` example ([here](https://github.com/Kitware/vtk-js/blob/master/Sources/Rendering/Core/ImageMapper/example/index.js#L36-L38), either with that interactorStyle or with the slice manipulator), I see that the z slice index goes between [-18; 216], even though the bounds of that image are [0; 200].

Thoughts? Also, does that only work for the z axis? We could reclamp the camera focalpoint by checking if the index is out of bounds in `setZSliceFromCamera`, but that doesn't seem clean to me.

Reply:
After discussion with @jourdain , we will leave this as is, and implement a `rangeManipulator` that will be generic enough to control scalar values (like slice, window, level), from horizontal drag, vertical drag, or pinch.